mmengine/_strategy/base.py

```diff
@@ -0,0 +1,906 @@
+            # Since `OptimWrapperDict` inherits from `OptimWrapper`,
+            # `isinstance(self.optim_wrapper, OptimWrapper)` cannot tell
+            # whether `self.optim_wrapper` is an `OptimizerWrapper` or
+            # `OptimWrapperDict` instance. Therefore, here we simply check
+            # self.optim_wrapper is not an `OptimWrapperDict` instance and
+            # then assert it is an OptimWrapper instance.
+            assert isinstance(optim_wrapper, BaseOptimWrapper), (
+                '`build_optimizer` should be called before'
+                '`build_param_scheduler` because the latter depends '
+                'on the former')
+            param_schedulers = self._build_param_scheduler(
+                scheduler, optim_wrapper, default_args)  # type: ignore
+            return param_schedulers
+        else:
+            param_schedulers = dict()
+            for name, optimizer in optim_wrapper.items():
+                if isinstance(scheduler, dict) and 'type' not in scheduler:
+                    # scheduler is a dict and each item is a ParamScheduler
+                    # object or a config to build ParamScheduler objects
+                    param_schedulers[name] = self._build_param_scheduler(
+                        scheduler[name], optimizer, default_args)
+                else:
+                    param_schedulers[name] = self._build_param_scheduler(
+                        scheduler, optimizer, default_args)
+
+            return param_schedulers
+
+    def build_logger(
+        self,
+        log_level: Union[int, str] = 'INFO',
+        log_file: Optional[str] = None,
+        **kwargs,
+    ) -> MMLogger:
+        """Build a global asscessable MMLogger.
+
+        Args:
+            log_level (int or str): The log level of MMLogger handlers.
+                Defaults to 'INFO'.
+            log_file (str, optional): Path of filename to save log.
+                Defaults to None.
+            **kwargs: Remaining parameters passed to ``MMLogger``.
+
+        Returns:
+            MMLogger: A MMLogger object build from ``logger``.
+        """
+        if log_file is None:
+            log_file = osp.join(self.log_dir, f'{self._timestamp}.log')
+
+        log_cfg = dict(log_level=log_level, log_file=log_file, **kwargs)
+        log_cfg.setdefault('name', self.experiment_name)
+        # `torch.compile` in PyTorch 2.0 could close all user defined handlers
+        # unexpectedly. Using file mode 'a' can help prevent abnormal
+        # termination of the FileHandler and ensure that the log file could
+        # be continuously updated during the lifespan of the runner.
+        log_cfg.setdefault('file_mode', 'a')
+
+        return MMLogger.get_instance(**log_cfg)  # type: ignore
+
+    def model_state_dict(self) -> dict:
+        """Get model state dict."""
+        from mmengine.runner import get_state_dict, weights_to_cpu
+        return weights_to_cpu(get_state_dict(self.model))
```

Comment:
https://github.com/open-mmlab/mmengine/blob/478c952a6d58c06715de890a337e185b68513fe4/mmengine/runner/runner.py#L2191C26-L2191C26

`get_state_dict` should be replaced by `model.state_dict`